lib/workload_mapper.rb

```diff
@@ -0,0 +1,87 @@
+# Copyright (c) 2013-2015 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of version 3 of the GNU General Public License as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE about this file by physical or electronic mail,
+# you may find current contact information at www.suse.com
+class WorkloadMapper
+  attr_reader :system, :workloads, :current_workload
+  def initialize(system_description, &block)
+    @system = system_description
+    @workloads = {}
+    @current_workload = nil
+    load_clues
+    instance_eval(&block) if block_given?
+  end
+
+  def write_compose_file(path)
+    compose_nodes = {}
+    workloads.each do |workload, config|
+      compose_nodes.merge!(compose_service(workload, config))
+    end
+    File.write(File.join(path, "docker-compose.yml"), compose_nodes.to_yaml)
+  end
+
+  def compose_service(workload, config)
+    name = config["service"]
+    service = {
+      name => compact(look_up_table[workload])
+    }
+    fill_in_template(service[name], config["parameters"])
+    service
+  end
+
+  def identify(name, service = nil)
+    @current_workload = name
+    workloads[name] = {
+      "service" => !!service ? service : name
+    }
+  end
+
+  def parameter(name, value)
+    workloads[current_workload]["parameters"] ||= Hash.new
+    workloads[current_workload]["parameters"][name] = value
+  end
+
+  def look_up_table
+    result = {}
+    files = Dir["#{File.expand_path(workload_mapper_path)}/*.yml"]
+    files.each do |file_path|
+      result.merge!(YAML::load(File.read(file_path)))
+    end
+    result
+  end
+
+  def load_clues
+    files = Dir["#{File.expand_path(workload_mapper_path)}/*.rb"]
+    files.each do |file_path|
+      instance_eval(File.read(file_path))
```

Comment:
I think I would separate the DSL from the mapper class, put the methods accessible from the clues into its own context.